envoy.dependency.cve_scan/envoy/dependency/cve_scan/abstract/cpe.py

```diff
@@ -0,0 +1,104 @@
+"""Abstract CPE."""
+
+import re
+
+import abstracts
+
+from envoy.base import utils
+
+from envoy.dependency.cve_scan.exceptions import CPEError
+from . import dependency
+
+
+FUZZY_DATE_RE = re.compile(r'(\d{4}).?(\d{2}).?(\d{2})')
+FUZZY_SEMVER_RE = re.compile(r'(\d+)[:\.\-_](\d+)[:\.\-_](\d+)')
+
+
+class ACPE(metaclass=abstracts.Abstraction):
+    """Model a subset of CPE fields that are used in CPE matching."""
+
+    @classmethod
+    def from_string(cls, cpe_str: str) -> "ACPE":
+        """Generate a CPE object from a CPE string."""
+        components = cpe_str.split(':')
+        if len(components) < 6 or not cpe_str.startswith('cpe:2.3:'):
+            raise CPEError(
+                f"CPE string ({cpe_str}) must be a valid CPE v2.3 string")
+        return cls(*components[2:6])
+
+    def __init__(
+            self,
+            part: str,
+            vendor: str,
+            product: str = "*",
+            version: str = "*") -> None:
+        self.part = part
+        self.vendor = vendor
+        self.product = product
+        self.version = version
+
+    def __str__(self):
+        return (
+            f"cpe:2.3:{self.part}:{self.vendor}:"
+            f"{self.product}:{self.version}")
+
+    @property
+    def vendor_normalized(self) -> str:
+        """Return a normalized CPE where only part and vendor are
+        significant."""
+        return str(self.__class__(self.part, self.vendor, '*', '*'))
+
+    def dependency_match(self, dep: dependency.ADependency) -> bool:
+        """Heuristically match dependency metadata against CPE."""
+
+        dep_cpe = self.__class__.from_string(
+            utils.typed(str, dep.cpe))
+
+        # We allow Envoy dependency CPEs to wildcard the 'product', this is
+        # useful for
+        # LLVM where multiple product need to be covered.
+        matching_parts = (
+            self.part == dep_cpe.part
+            and self.vendor == dep_cpe.vendor
+            and (
+                dep_cpe.product == '*'
+                or self.product == dep_cpe.product))
+        if not matching_parts:
+            return False
+
+        # TODO: fix/remove remaining - this always evaluates to True
+        version_match = (
+            # Wildcard versions always match.
+            self.version == '*'
+            # An exact version match is a hit.
+            or self.version == dep.version
+            # Allow the 'release_date' dependency metadata to substitute
+            # for date.
+            # TODO(htuch): Consider fuzzier date ranges.
+            or self.version == dep.release_date)
+        if version_match:
+            return True
+
+        # Try a fuzzy date match to deal with versions like fips-20190304 in
```

Comment:
@htuch this is my other concern

not sure how this code was hit in the past but afaict in the current implementation (that this is replacing) it never gets hit

i believe this is because none of the cpes listed in repository_locations.bzl are versioned - cf:

```console
$ git grep cpe bazel/repository_locations.bzl | grep -v "N/A"
```

Im wondering whether we want to keep given its not used

Reply:
It wasn't the repository_locationz.bzl CPE that was versioned, rather the version field in something like https://github.com/envoyproxy/envoy/blob/aac4631c26745de1a39556abc68bd1d3ed95aa2f/bazel/repository_locations.bzl#L97 was not a real "version" which could be used to compare against the CPE version in the NIST database. Hence this.

Reply:
ok, ill leave it there for now, its not tested, but we can revisit - i want to add some more tests after anyway